### PR TITLE
refactor: Add header types

### LINKED ISSUE #13587: Add TypeScript types to HeaderGlobalAction

### PR DESCRIPTION
**Pull request by danoro96**
Closes #13587, #13588, #13589, #13585

Add types for Header, HeaderGlobalBar, HeaderMenu, HeaderGlobalAction
#### Changelog

**Changed**

- Added Header.tsx
- Added HeaderGlobalBar.tsx
- Added HeaderMenu.tsx
- Added HeaderGlobalAction.tsx

#### Testing / Reviewing

Fire up StoryBook and check that everything works as expected

### CODE REVIEW

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Thanks for taking this on! I think it looks great. The only thing I see is that each interface needs the prop comments copied from the proptypes for future documentation efforts.

File: packages/react/src/components/UIShell/HeaderGlobalAction.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+   */
+  'aria-label'?: string;
+  /**
+   * Required props for the accessibility label of the button
+   */
+  'aria-labelledby'?: string;
+  /**
+   * Provide a custom icon for this global action
+   */
+  children?: ReactNode;
```
Comment (guidari) [inline packages/react/src/components/UIShell/HeaderGlobalAction.tsx]: This one should be required
```suggestion
  children: ReactNode;
```

**Review by guidari — APPROVED** [review]
LGTM! 🚀 @danoro96 
Just one minor change

**Comment by danoro96** [comment]
Hey all,

It says that there is a merge conflict with `HeaderMenu.js`.
I pulled main and tried to rebase but git says I'm up to date.
Not sure how to fix a merge conflict with a file that is no longer there. 

Any ideas?

**Comment by danoro96** [comment]
> Hey all,
> 
> It says that there is a merge conflict with `HeaderMenu.js`. I pulled main and tried to rebase but git says I'm up to date. Not sure how to fix a merge conflict with a file that is no longer there.
> 
> Any ideas?

I was able to remove and re-add the file. That fixed the issue.

File: packages/react/src/components/UIShell/HeaderMenu.tsx [hunk 2]
```diff
@@ -201,25 +264,23 @@ class HeaderMenu extends React.Component {
...
-          child.props.isActive ||
-          child.props.isCurrentPage ||
-          (child.props.children instanceof Array &&
-            hasActiveDescendant(child.props.children))
-      );
+    const hasActiveChildren = React.Children.toArray(children).some((child) => {
+      if (React.isValidElement(child)) {
+        child.props.isActive || child.props.isCurrentPage;
+      }
+    });
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/HeaderMenu.tsx]: This is a functional change, right? This no longer recursively looks through descendant children. I think we'll want to retain that.

**Review by tay1orjones — CHANGES_REQUESTED** [review]

**Pull request closed without merging**